### ISSUE TITLE
Tests failing on 7.0.0-rc.3

### ISSUE THREAD

**Issue by ronaldmansveld** [angular/angular-cli#12620]
### Bug Report or Feature Request (mark with an `x`)
```
- [X] bug report -> please search issues before submitting
- [ ] feature request
```

### Command (mark with an `x`)
```
- [ ] new
- [ ] build
- [ ] serve
- [X] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Versions
Windows 10
NPM 6.2.0

Angular CLI: 7.0.0-rc.3
Node: 10.8.0
OS: win32 x64
Angular: 7.0.0-rc.1
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router, service-worker

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.9.0-rc.3
@angular-devkit/build-angular      0.9.0-rc.3
@angular-devkit/build-optimizer    0.9.0-rc.3
@angular-devkit/build-webpack      0.9.0-rc.3
@angular-devkit/core               7.0.0-rc.3
@angular-devkit/schematics         7.0.0-rc.3
@angular/cdk                       7.0.0-rc.2
@angular/cli                       7.0.0-rc.3
@angular/material                  7.0.0-rc.2
@angular/material-moment-adapter   7.0.0-rc.2
@angular/pwa                       0.8.5
@ngtools/webpack                   7.0.0-rc.3
@schematics/angular                7.0.0-rc.3
@schematics/update                 0.9.0-rc.3
rxjs                               6.3.3
typescript                         3.1.3
webpack                            4.19.1

### Repro steps
1) Create new project
2) Run test
3) Add service
4) Start added tests to spec file

### The log given by the failure
ERROR in ./src/app/common-services/services/language/language.service.ts
Module build failed (from ./node_modules/@ngtools/webpack/src/index.js):
Error: C:\dev\dotnet\Recon\ResponsiveWeb\v3\src\app\common-services\services\language\language.service.ts is missing from the TypeScript compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
    at AngularCompilerPlugin.getCompiledFile (C:\dev\dotnet\Recon\ResponsiveWeb\v3\node_modules\@ngtools\webpack\src\packages\ngtools\webpack\src\angular_compiler_plugin.ts:1021:15)
    at plugin.done.then (C:\dev\dotnet\Recon\ResponsiveWeb\v3\node_modules\@ngtools\webpack\src\packages\ngtools\webpack\src\loader.ts:49:29)
    at process._tickCallback (internal/process/next_tick.js:68:7)
 @ ./src/app/common-services/services/language/language.service.spec.ts 1:0-53 5:22-37
 @ ./src sync \.spec\.ts$
 @ ./src/test.ts

### Desired functionality
Test should not randomly fail after adding a test case

### Mention any other details that might be useful
Using custom schematics that adds an extra directory level under a module

**Comment by ronaldmansveld** [comment]
Some extra information:

- I tried adding '**/*.ts' to the include array of tsconfig.spec.json, with no effect
- Sometimes it errors on the service not being part of the typescript compilation, sometimes it errors on the containing module not being part, and sometimes the services is shown as a warning (with the same message) and the module as an error. Haven't found a way to replicate either of those consistently
- Ctrl-C-ing the failing run, and restarting ng test sometimes runs the tests succesfully, sometimes it just hangs and runs no tests at all

**Comment by ronaldmansveld** [comment]
Managed to narrow the issue further down:

After some changes in my spec file (as that had errors that were not reported anywhere apart from the console in the browser running karma), the issue now only happens when changing and saving the service. Changing and saving the spec file re-runs the tests normally without issues.

Saving the service while ng test is still running makes ng test crash with the 'is missing from TypeScript compilation' error.

Both restarting ng test, as well as reloading the browser window running karma will execute the tests normally.

**Comment by alan-agius4** [comment]
This seems like a bug but we'll need to look at a reproduction to find and fix the problem. Can you setup a minimal repro please?

You can read [here](https://github.com/angular/angular-cli/blob/master/CONTRIBUTING.md#-submitting-an-issue) why this is needed. A good way to make a minimal repro is to create a new app via `ng new repro-app` and adding the minimum possible code to show the problem. Then you can push this repository to github and link it here.

This might be related to your directory structure so its really important to get an accurate repro to diagnose this.

![image](https://user-images.githubusercontent.com/17563226/47137887-2e747680-d2b8-11e8-9adc-c9a511d3e24a.png)

**Comment by ronaldmansveld** [comment]
Hi @alan-agius4 ,

I'll start creating a repo now, and will post a link here with the steps taken to create the repo, to pinpoint the step where it breaks down.

Ronald

**Comment by ronaldmansveld** [comment]
Hi @alan-agius4 

The repository with a reproduction can be found here: https://github.com/ronaldmansveld/ng-cli-12620

For completion, here are the steps I've taken to create this repo, as well as points where the tests don't run as expected:

### Create app
npm init
npm install @angular/cli@next
ng new repro-app
cd repro-app/
### Add module and service (note extra directory level)
ng g module test
ng g service test/services/test --flat=false --module=test
ng test -> runs fine
### Start development and check if tests run
Refactor test.spec.ts to be isolated (not using testbed) -> *tests run fine*
Add spec without corresponding implementation -> *tests run fine (1 test fails, as expected)*
Add implementation to service -> *karma reloads in browser, but no tests are run. No error on CLI*
Save service again without changes -> *Service tests don't run (only app does)! (But no error on CLI)*
Stop ng test
Restart ng test -> *All tests run*
Refactor service (add _items) and save file -> *Tests don't run*
Add new test -> *tests don't run, 'missing from typescript compilation' error + 'addItem does not exist on service' error* (2)
Add implementation of addItem and save service -> *tests don't run*
Reload browser window running karma -> *tests don't run*
Stop ng test
Restart ng test -> *tests run*
Save service without changes -> *browser window with karma reloads, but no tests run*
Save spec.ts without changes -> *'test.service.ts is missing from the typescript compilation' error* (1)


(1) Output on command line from last error:
---------------------------------------
× ｢wdm｣: Hash: dea2651a18dd2b11e27d
Version: webpack 4.19.1
Time: 114ms
Built at: 10/18/2018 10:55:37 AM
       Asset      Size     Chunks         Chunk Names
     main.js  37.7 KiB       main         main
polyfills.js   525 KiB  polyfills  [big]  polyfills
   styles.js    20 KiB     styles         styles
   vendor.js  7.51 MiB     vendor  [big]  vendor
Entrypoint main = vendor.js main.js
Entrypoint polyfills = polyfills.js
Entrypoint styles = vendor.js styles.js
[./node_modules/@angular-devkit/build-angular/src/angular-cli-files/models/jit-polyfills.js] 233 bytes {polyfills}
[./node_modules/@angular/compiler/fesm5/compiler.js] 1.11 MiB {vendor}
[./node_modules/@angular/compiler/fesm5/testing.js] 11.4 KiB {vendor}
[./node_modules/@angular/core/fesm5/core.js] 817 KiB {vendor}
[./node_modules/@angular/core/fesm5/testing.js] 76.3 KiB {vendor}
[./node_modules/@angular/platform-browser-dynamic/fesm5/testing.js] 9.68 KiB {vendor}
[./node_modules/@angular/platform-browser/fesm5/testing.js] 7.12 KiB {vendor}
[./node_modules/tslib/tslib.es6.js] 8.83 KiB {vendor}
[./node_modules/zone.js/dist/zone-testing.js] 67.4 KiB {vendor}
[./src sync recursive \.spec\.ts$] ./src sync \.spec\.ts$ 242 bytes {main} [built]
[./src/polyfills.ts] 3.14 KiB {polyfills}
[./src/styles.css] 1.38 KiB {styles}
[./src/test.ts] 599 bytes {main}
[0] multi ./src/polyfills.ts ./node_modules/@angular-devkit/build-angular/src/angular-cli-files/models/jit-polyfills.js 40 bytes {polyfills}
[1] multi ./src/styles.css 28 bytes {styles}
    + 296 hidden modules

ERROR in ./src/app/test/services/test/test.service.ts
Module build failed (from ./node_modules/@ngtools/webpack/src/index.js):
Error: C:\dev\ronaldmansveld\ng-cli-12620\repro-app\src\app\test\services\test\test.service.ts is missing from the TypeScript compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
    at AngularCompilerPlugin.getCompiledFile (C:\dev\ronaldmansveld\ng-cli-12620\repro-app\node_modules\@ngtools\webpack\src\packages\ngtools\webpack\src\angular_compiler_plugin.ts:1021:15)
    at plugin.done.then (C:\dev\ronaldmansveld\ng-cli-12620\repro-app\node_modules\@ngtools\webpack\src\packages\ngtools\webpack\src\loader.ts:49:29)
    at process._tickCallback (internal/process/next_tick.js:68:7)
 @ ./src/app/test/services/test/test.service.spec.ts 1:0-45 5:22-33
 @ ./src sync \.spec\.ts$
 @ ./src/test.ts
---------------------------------------


(2) Output from error:
---------------------------------------
× ｢wdm｣: Hash: dea2651a18dd2b11e27d
Version: webpack 4.19.1
Time: 92ms
Built at: 10/18/2018 10:52:28 AM
       Asset      Size     Chunks         Chunk Names
     main.js  37.7 KiB       main         main
polyfills.js   525 KiB  polyfills  [big]  polyfills
   styles.js    20 KiB     styles         styles
   vendor.js  7.51 MiB     vendor  [big]  vendor
Entrypoint main = vendor.js main.js
Entrypoint polyfills = polyfills.js
Entrypoint styles = vendor.js styles.js
[./node_modules/@angular-devkit/build-angular/src/angular-cli-files/models/jit-polyfills.js] 233 bytes {polyfills}
[./node_modules/@angular/compiler/fesm5/compiler.js] 1.11 MiB {vendor}
[./node_modules/@angular/compiler/fesm5/testing.js] 11.4 KiB {vendor}
[./node_modules/@angular/core/fesm5/core.js] 817 KiB {vendor}
[./node_modules/@angular/core/fesm5/testing.js] 76.3 KiB {vendor}
[./node_modules/@angular/platform-browser-dynamic/fesm5/testing.js] 9.68 KiB {vendor}
[./node_modules/@angular/platform-browser/fesm5/testing.js] 7.12 KiB {vendor}
[./node_modules/tslib/tslib.es6.js] 8.83 KiB {vendor}
[./node_modules/zone.js/dist/zone-testing.js] 67.4 KiB {vendor}
[./src sync recursive \.spec\.ts$] ./src sync \.spec\.ts$ 242 bytes {main} [built]
[./src/polyfills.ts] 3.14 KiB {polyfills}
[./src/styles.css] 1.38 KiB {styles}
[./src/test.ts] 599 bytes {main}
[0] multi ./src/polyfills.ts ./node_modules/@angular-devkit/build-angular/src/angular-cli-files/models/jit-polyfills.js 40 bytes {polyfills}
[1] multi ./src/styles.css 28 bytes {styles}
    + 296 hidden modules

ERROR in ./src/app/test/services/test/test.service.ts
Module build failed (from ./node_modules/@ngtools/webpack/src/index.js):
Error: C:\dev\ronaldmansveld\ng-cli-12620\repro-app\src\app\test\services\test\test.service.ts is missing from the TypeScript compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
    at AngularCompilerPlugin.getCompiledFile (C:\dev\ronaldmansveld\ng-cli-12620\repro-app\node_modules\@ngtools\webpack\src\packages\ngtools\webpack\src\angular_compiler_plugin.ts:1021:15)
    at plugin.done.then (C:\dev\ronaldmansveld\ng-cli-12620\repro-app\node_modules\@ngtools\webpack\src\packages\ngtools\webpack\src\loader.ts:49:29)
    at process._tickCallback (internal/process/next_tick.js:68:7)
 @ ./src/app/test/services/test/test.service.spec.ts 1:0-45 5:22-33
 @ ./src sync \.spec\.ts$
 @ ./src/test.ts
ERROR in src/app/test/services/test/test.service.spec.ts(19,13): error TS2339: Property 'addItem' does not exist on type 'TestService'.
---------------------------------------

**Comment by waybackrider** [comment]
can confirm this!
Before the Error occurs, Tests are reported successful which are definitley not because of a code change, so you can rely on ng test only on the first run!

**Comment by filipesilva** [comment]
I think this is the same as https://github.com/angular/angular-cli/issues/12657 and was fixed by https://github.com/angular/angular-cli/pull/12591, released in `@angular-devkit/build-angular@0.10.3`.
